### PR TITLE
docs: add more thorough examples for predicates

### PR DESCRIPTION
**Pull request by maschad**
This PR seeks to improve the documentation surrounding predicates in the SDK, this includes

- documentation of the [Predicate methods](https://github.com/FuelLabs/fuels-ts/pull/2345/files#diff-8f528b24de2456d5421a8cd534e5d9d63d73c009665a62461f5c9565aa6b72ea)
- A walkthrough [example](https://github.com/FuelLabs/fuels-ts/pull/2345/files#diff-69098d51c0733a29214f157d838eb613e8ba64ea594b1e2e526b0beac1197435) guide of how one would modify the existing [create fuels predicate example](https://github.com/FuelLabs/fuels-ts/blob/master/templates/nextjs/src/pages/predicate.tsx) to use a configurable predicate.

Closes https://github.com/FuelLabs/fuels-ts/issues/2311

### CODE REVIEW

File: apps/docs/src/guide/predicates/methods.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Interacting With Predicates
+
+There are 4 ways to interact with predicates: `populateTransactionPredicateData`, `sendTransaction`, `simulateTransaction`, `getResourcesToSpend`
+
+
+## `populateTransactionPredicateData`
```
Comment (danielbate) [inline apps/docs/src/guide/predicates/methods.md:6]: This isn't strictly true as a `Predicate` extends an `Account` so it inherits all the account based functionality (submitting, querying etc).

Maybe this could be made more general with the subheadings:
- Checking Balances - `getBalances` & `getResourcesToSpend` (make this one tx specific)
- Transactions -`sendTransaction` / `simulateTransaction` and `populateTransactionPredicateData` (for where the predicate is added to a tx builder) 
- Transfers - `createTransfer` / `transfer`

FYI I don't necessarily agree with this design choice as there are some things in there that don't make sense i.e. signing and the use of connectors, but we should highlight the account based functionality that it does use.

File: apps/docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+import {
+  DocSnippetProjectsEnum,
+  getDocsSnippetsForcProject,
+} from '../../../test/fixtures/forc-projects';
+import { getTestWallet } from '../../utils';
+
+/**
+ * @group node
+ */
+
```
Comment (danielbate) [inline apps/docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts:14]:
```suggestion
```

File: apps/docs/src/guide/creating-a-fuel-dapp/index.md [hunk 3]
```diff
@@ -185,8 +185,11 @@ Whenever you want to add a new feature to your dApp and quickly prototype things
 
 - Now that you have a basic counter dApp running and have the `npm create fuels` workflow powering you, you can start building more complex dApps using the Fuel Stack. A good place to start for ideas and reference code is the [Sway Applications Repo](https://github.com/FuelLabs/sway-applications).
 
+- As you may have noticed, there are different types of programs in your dApp, feel free to explore [Predicates](https://docs.fuel.network/docs/fuels-ts/predicates/) and [Scripts](https://docs.fuel.network/docs/fuels-ts/scripts/), which are both important differentiators in the Fuel Stack.
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/creating-a-fuel-dapp/index.md:188]: Maybe I misinterpreted your suggestion, perhaps you should outline it in a separate issue.

[the remark this answers is not in the document]
Comment (danielbate) [reply apps/docs/src/guide/creating-a-fuel-dapp/index.md:188]: To get a consensus, do we think this is enough of an extension, or should we be giving specific code examples for extending the demo with scripts/predicates?
> cc @FuelLabs/sdk-ts

File: apps/docs/src/guide/predicates/methods.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
+# Interacting With Predicates
+
+The `Predicate` class extends the [`Account`](https://docs.fuel.network/docs/fuels-ts/account/) class, and therefore it inherits all of it's methods.  Therefore, there are multiple ways to interact with predicates, but broadly speaking there are 3 general use cases that would be of interest to a user. We can think of them in terms of `Checking Balances`, `Transactions` and `Transfers`
```
Comment (arboleya) [inline apps/docs/src/guide/predicates/methods.md:3]:
```suggestion
The `Predicate` class extends the [`Account`](https://docs.fuel.network/docs/fuels-ts/account/) class, inheriting all its methods. Therefore, there are multiple ways to interact with predicates, but broadly speaking, we can think about three:
 - `Checking Balances`
 - `Transactions`
 - `Transfers`
```

File: apps/docs/src/guide/predicates/methods.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+### `getBalances`
+
+This will return the balances of the wallet that is being used to interact with the predicate.
+
+<<< @/../../docs-snippets/src/guide/wallets/checking-balances.test.ts#checking-balances-2{ts:line-numbers}
+
+### `getResourcesToSpend`
+
+This will return the resources that will be spent when calling the predicate. This is useful when creating the predicate transaction so that you can ensure that the wallet has sufficient funds to cover the transaction fees.
```
Comment (arboleya) [inline apps/docs/src/guide/predicates/methods.md:16]:
```suggestion
This will return the resources that are spent when calling the predicate. This is useful when creating the predicate transaction to ensure that the wallet has sufficient funds to cover the transaction fees.
```

File: apps/docs/src/guide/predicates/methods.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-1{ts:line-numbers}
+
+## Transactions
+
+## `sendTransaction`
+
+Once you've populated the transaction with the populated data you can send the transaction to the node using the `sendTransaction` method.
```
Comment (arboleya) [inline apps/docs/src/guide/predicates/methods.md:24]:
```suggestion
Once you've populated the transaction with the populated data, you can send it to the node using the `sendTransaction` method.
```

File: apps/docs/src/guide/predicates/methods.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+## `simulateTransaction`
+
+The `simulateTransaction` method should be used to dry-run a predicate call, ensuring that the wallet used has sufficient funds to cover the transaction fees, without consuming any resources.
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-3{ts:line-numbers}
+
+### `populateTransactionPredicateData`
+
+Once you've instantiated a `new Predicate` you will want to create a transaction request containing the `gasLimit` and `maxFee` then pass this to the `populateTransactionPredicateData` method which will populate the transaction with the required data to be sent to the node.
```
Comment (arboleya) [inline apps/docs/src/guide/predicates/methods.md:36]:
```suggestion
Once you've instantiated a `new Predicate` you will want to create a transaction request containing the `gasLimit` and `maxFee` and then pass this to the `populateTransactionPredicateData` method, which will populate the transaction with the required data to be sent to the node.
```

**Review by arboleya — COMMENTED** [review]
A couple of grammar suggestions, tyops, and nits.

Feel free to choose the ones you like and discard the rest.

**Review by Torres-ssf — DISMISSED** [review]
Good job 👍

File: apps/docs/src/guide/creating-a-fuel-dapp/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/creating-a-fuel-dapp/index.md:188]: I'm not sure what that lite version would entail @petertonysmith94 as there isn't anything analogous existing in documentation and _create fuels_ already offers predicate and scripts by default. I think it may need to have a a separate discussion / issue (rather than doing it in this thread) to come to consensus on how we would like to extend the docs.

File: apps/docs/src/guide/predicates/methods.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+### `getResourcesToSpend`
+
+This will return the resources that will be spent when calling the predicate. This is useful when creating the predicate transaction so that you can ensure that the wallet has sufficient funds to cover the transaction fees.
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-1{ts:line-numbers}
+
+## Transactions
+
+## `sendTransaction`
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md:22]: I intentionally made these sub-headings of the broader headings of  `Checking Balances`, `Transactions` and `Transfers` as I think that categorizes these methods more precisely

File: apps/docs/src/guide/predicates/methods.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+Once you've populated the transaction with the populated data you can send the transaction to the node using the `sendTransaction` method.
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-2{ts:line-numbers}
+
+## `simulateTransaction`
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md:28]: See https://github.com/FuelLabs/fuels-ts/pull/2345#discussion_r1609613466

File: apps/docs/src/guide/predicates/methods.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+Once you've instantiated a `new Predicate` you will want to create a transaction request containing the `gasLimit` and `maxFee` then pass this to the `populateTransactionPredicateData` method which will populate the transaction with the required data to be sent to the node.
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-4{ts:line-numbers}
+
+## Transfers
+
+### `createTransfer`
+
+The `createTransfer` method is used to create a transfer object that can be used to send funds to another wallet. This can be useful should you need to modify the transfer object or use it's properties before sending it to the node.
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: ee69754

File: apps/docs/src/guide/predicates/methods.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+<<< @/../../docs-snippets/src/guide/predicates/send-and-spend-funds-from-predicates.test.ts#send-and-spend-funds-from-predicates-8{ts:line-numbers}
+
+### `transfer`
+
+Once you've created a `transfer` object using the `createTransfer` method can send the funds to another wallet using the `transfer` method.
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: https://github.com/FuelLabs/fuels-ts/commit/ee69754a2d64947dcb59dc454b35a6cbeed52108

File: apps/docs/src/guide/predicates/methods.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: https://github.com/FuelLabs/fuels-ts/commit/ee69754a2d64947dcb59dc454b35a6cbeed52108

File: apps/docs/src/guide/predicates/methods.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md:29]: My bad, I misread your suggestion 😅  I agree with you, thanks for pointing this out.

File: apps/docs/src/guide/creating-a-fuel-dapp/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/creating-a-fuel-dapp/index.md:188]: A picture is worth a thousand words ✍🏾  🌹  thanks for clarifying.

File: apps/create-fuels-counter-guide/src/pages/predicate.tsx [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+      return toast.error(
+        "Your wallet does not have enough funds. Please click the 'Top-up Wallet' button in the top right corner, or use the local faucet."
+      );
+    }
+
+    if (!pin) {
+      return toast.error("Please enter a pin");
+    }
+
+    const configurable = { PIN: bn(pin) };
```
Comment (maschad) [inline apps/create-fuels-counter-guide/src/pages/predicate.tsx:107]: @Dhaiwat10 I'm uncertain as to whether this is idiomatic React, I primarily did things this way as it seemed to be more favourable to the nuances of how React manages state. I'd be keen to hear your insights.

File: apps/docs/src/guide/creating-a-fuel-dapp/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/creating-a-fuel-dapp/index.md:206]: 3ab9c17eb3ef9d52347313981422a6d12e63f2fa

File: apps/docs-snippets/test/fixtures/forc-projects/index.ts [hunk 13]
```diff
@@ -34,6 +34,7 @@ export enum DocSnippetProjectsEnum {
   SCRIPT_SIGNING = 'script-signing',
   INPUT_OUTPUT_TYPES = 'input-output-types',
   BYTECODE_INPUT = 'bytecode-input',
+  CONFIGURABLE_PIN = 'configurable-pin',
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs-snippets/test/fixtures/forc-projects/index.ts:37]: This should actually be removed, good catch

File: apps/docs/.vitepress/config.ts [hunk 14]
```diff
@@ -83,6 +83,10 @@ export default defineConfig({
                 text: 'Deploying a dApp to Testnet',
                 link: '/guide/creating-a-fuel-dapp/deploying-a-dapp-to-testnet',
               },
+              {
+                text: 'Creating a Fuel dApp with Predicates',
+                link: '/guide/creating-a-fuel-dapp/creating-a-fuel-dapp-with-predicates',
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/.vitepress/config.ts:88]: Good point

File: apps/docs/src/guide/creating-a-fuel-dapp/creating-a-fuel-dapp-with-predicates.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+<<< @/../../docs-snippets/test/fixtures/forc-projects/configurable-pin/src/main.sw#predicate-with-configurable-pin-1{rust:line-numbers}
+
+2. Modifying the Frontend
+
+We will now a new button to the frontend that will update the `pin` in the predicate when clicked. To do this, we will modify the `./src/pages/predicate.tsx` file.
+
+Firstly, we will add a function called `changePin`, which will use the current pin in state to update the pin in the predicate as well as transfer 1000 to the predicate.
+
+<<< @/../../create-fuels-counter-guide/src/pages/predicate.tsx#change-pin-react-function{ts:line-numbers}
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/creating-a-fuel-dapp/creating-a-fuel-dapp-with-predicates.md:24]: Correct, these changes actually should have only remained in the [creating fuels counter guide example](https://github.com/FuelLabs/fuels-ts/pull/2345/files#diff-949bdc64dfe5fd74abb1df40797406fccda928b3dafa568c091fe61019cfb4a2) but I think they were accidentally placed here as well.

**Review by nedsalk — COMMENTED** [review]
Hmm, I commented ages ago on this PR but I think the comments didn't show up because I didn't _submit_ the review :/

File: apps/docs/src/guide/predicates/methods.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+ - `Transfers`
+
+
+## Checking Balances
+
+### `getBalances`
+
+This will return the balances of the wallet that is being used to interact with the predicate.
+
+<<< @/../../docs-snippets/src/guide/wallets/checking-balances.test.ts#checking-balances-2{ts:line-numbers}
```
Comment (nedsalk) [inline apps/docs/src/guide/predicates/methods.md:15]: The snippet is in `wallets/...` and it should be in `predicates/...`.

Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md:15]: It's being reused for this purpose but it was originally uses by the wallets documentation.

File: apps/docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    // #region interacting-with-predicates-4
+    let transactionRequest = new ScriptTransactionRequest({ gasLimit: 2000, maxFee: bn(0) });
+    transactionRequest.addCoinOutput(receiver.address, 100, baseAssetId);
+
+    const predicateResources = await provider.getResourcesToSpend(predicate.address, quantity);
+
+    transactionRequest.addResources(predicateResources);
+
+    transactionRequest = predicate.populateTransactionPredicateData(transactionRequest);
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts:134]: That's a good point,  I had added this before. Should we deprecate this method? I can't think of a use case for it otherwise. cc @Torres-ssf

File: apps/docs/src/guide/predicates/methods.md (hunk 16 above)
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md:15]: https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a

File: apps/docs/src/guide/predicates/methods.md [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+### `sendTransaction`
+
+Once you've populated the transaction with the populated data, you can send it to the node using the `sendTransaction` method.
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-2{ts:line-numbers}
+
+### `simulateTransaction`
+
+The `simulateTransaction` method should be used to dry-run a predicate call, ensuring that the wallet used has sufficient funds to cover the transaction fees, without consuming any resources.
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a

File: apps/docs/src/guide/predicates/methods.md [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+You may want to use this method when using a predicate in an existing transaction request.
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-1{ts:line-numbers}
+
+## Transactions
+
+### `sendTransaction`
+
+Once you've populated the transaction with the populated data, you can send it to the node using the `sendTransaction` method.
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a

File: apps/docs/src/guide/predicates/methods.md [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+This will return the balances of the wallet that is being used to interact with the predicate.
+
+<<< @/../../docs-snippets/src/guide/wallets/checking-balances.test.ts#checking-balances-2{ts:line-numbers}
+
+### `getResourcesToSpend`
+
+This will return the resources owned by a predicate so that they can be added to a transaction request.
+
+This method is called under the hood when using `transfer` or `createTransfer`.
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: [`80b13c2` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a)

File: apps/docs/src/guide/predicates/methods.md [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+ - `Checking Balances`
+ - `Transactions`
+ - `Transfers`
+
+
+## Checking Balances
+
+### `getBalances`
+
+This will return the balances of the wallet that is being used to interact with the predicate.
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: [`80b13c2` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a)

File: apps/docs/src/guide/predicates/methods.md [hunk 22]
```diff
@@ -10,15 +10,15 @@ The `Predicate` class extends the [`Account`](https://docs.fuel.network/docs/fue
 
 ### `getBalances`
 
-This will return the balances of the wallet that is being used to interact with the predicate.
+This will return the balances of the account that owns the resources in the predicate.
```
Comment (nedsalk) [inline apps/docs/src/guide/predicates/methods.md:13]:
```suggestion
This will return the balances of all assets owned by the predicate.
```
The predicate itself owns the resources and _it_ is the account.

File: apps/docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (Torres-ssf) [reply apps/docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts:134]: @maschad This method cannot be deprecated as it's still used by the `EVMWalletConnector`. Given the context of how this connector works, the predicate that it uses is only instantiated after its `sendTransaction` method is called with the TX already funded with predicate resources that were fetched as a normal account. So I suppose this connector will always use this method.

We also [use this method](https://github.com/FuelLabs/fuels-ts/blob/977d58ba50c55de64621c8d509e660937cb99991/packages/account/src/providers/provider.ts#L1141-L1143) within the TS SDK at `Provider.getTransactionCost`, to populate the fake resources we create for a dry-run, as they may belong to a Predicate. The issue https://github.com/FuelLabs/fuels-ts/issues/2362 addresses this problem

In your example, the Predicate instance is being used to fetch its resources, so we do not need to call `populateTransactionPredicateData`

[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts:134]: Thanks for the clarification, I suppose it may not make sense to expose this method in our public API given most consumers of the SDK would not have much use for it then.

File: apps/docs/src/guide/creating-a-fuel-dapp/working-with-predicates.md [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        value={pin as string}
+        onChange={(e) => setPin(e.target.value)}
+        placeholder="Enter a new pin"
+      />
+```
+
+Thirdly, we will add a button that calls the `changePin` function when clicked.
+
+```tsx
+  <Button onClick={changePin}>Change Pin</Button>
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/creating-a-fuel-dapp/working-with-predicates.md:40]: See https://github.com/FuelLabs/fuels-ts/pull/2345#discussion_r1647909789

File: apps/docs/src/guide/creating-a-fuel-dapp/working-with-predicates.md [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+Congratulations! That's all. You should now be able to see the modified predicate dApp running at `http://localhost:3000` with our newly added change pin functionality.
+
+You can find the complete source code of the dApp we built [here](https://github.com/FuelLabs/fuels-ts/tree/master/apps/create-fuels-counter-guide).
+
+## Next Steps
+
+- Now that you have a predicate dApp running and have the `npm create fuels` workflow powering you, you can start building more complex dApps using the Fuel Stack. A good place to start for ideas and reference code is the [Sway Applications Repo](https://github.com/FuelLabs/sway-applications).
+
+- If you want to deploy your dApp to the testnet, check out our [Deploying a dApp to Testnet](./deploying-a-dapp-to-testnet.md) guide.
+
```
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/guide/creating-a-fuel-dapp/working-with-predicates.md:52]: Good point, I suppose anyone who would want to deploy a contract would have looked at the previous guide

File: templates/nextjs/sway-programs/predicate/src/main.sw
[the remark this answers is not in the document]
Comment (maschad) [reply templates/nextjs/sway-programs/predicate/src/main.sw:1]: This change should actually be in the [create-fuels-demo-app](https://github.com/FuelLabs/fuels-ts/tree/mc/docs/improve-predicate-docs/apps/create-fuels-counter-guide/sway-programs) as a reference, so good catch.

File: apps/docs/src/guide/predicates/methods.md [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-3{ts:line-numbers}
+
+## Transfers
+
+### `createTransfer`
+
+The `createTransfer` method creates a transfer request and add the transfer details to that request. This method estimates the transaction cost via a dry-run, and then funds it with predicate resources, which is useful since the returned transaction request can now be submitted with greater certainty that it will succeed.
+
+ This can be useful should you need to modify the transfer request or use its properties before sending it to the node.
```
Comment (arboleya) [inline apps/docs/src/guide/predicates/methods.md:47]:
```suggestion
The `createTransfer` method creates a transaction request with all the necessary transfer details. It automatically estimates the transaction costs via a dry-run call and funds the request with the required predicate resources. After this, one can submit the returned transaction request with greater certainty that it will succeed.

However, please remember that you can still modify the transfer request details and use its properties before submitting it to the node.
```

I'd also say these two lines seem irrelevant to the purpose of the code snippet and could be deleted:

```ts
const chainId = provider.getChainId();

const txId = transactionRequest.getTransactionId(chainId);
```

File: templates/nextjs/sway-programs/predicate/src/main.sw
[the remark this answers is not in the document]
Comment (maschad) [reply templates/nextjs/sway-programs/predicate/src/main.sw:1]: [`b45bd9f` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/b45bd9f82824f6ad92952b4fdaea48ebf52fb98e)

File: apps/docs/src/public/working-with-predicates-end-result.png
[the remark this answers is not in the document]
Comment (maschad) [reply apps/docs/src/public/working-with-predicates-end-result.png:1]: [`b45bd9f` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/b45bd9f82824f6ad92952b4fdaea48ebf52fb98e)

File: apps/docs/src/guide/predicates/methods.md (hunk 16 above)
Comment (arboleya) [reply apps/docs/src/guide/predicates/methods.md:15]: Un-resolving this just to ask: What was the resolution?

File: apps/docs/src/guide/predicates/methods.md (hunk 25 above)
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: My apologies, when I applied the suggestion you made [here](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/e4e78bb58cd57c741b85d1273d7960aa0f2d3a40) Github automatically resolved the issue, and so that consideration eluded me. 

This code snippet is re-used from another [guide](https://github.com/FuelLabs/fuels-ts/blob/mc/docs/improve-predicate-docs/apps/docs/src/guide/predicates/send-and-spend-funds-from-predicates.md#L45), but I agree with your point, I'll implement a new snippet for this.

Comment (arboleya) [reply apps/docs/src/guide/predicates/methods.md]: Perhaps you can just move the two lines below where the snippet ends.

File: apps/docs/src/guide/predicates/methods.md (hunk 16 above)
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md:15]: I must have overlooked this comment, my bad. I agree with your suggestion.

File: apps/docs/src/guide/predicates/methods.md (hunk 25 above)
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: See [`2a8de44` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/2a8de44dc130234cfff5cc6b455c9ad6f1be188f)

File: apps/docs/src/guide/predicates/methods.md (hunk 16 above)
Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: See [`2a8de44` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/2a8de44dc130234cfff5cc6b455c9ad6f1be188f)

File: apps/docs/src/guide/predicates/methods.md (hunk 25 above)
Comment (arboleya) [reply apps/docs/src/guide/predicates/methods.md]: TBH, the text preceding this snippet could be clearer as well. 🙂 

We could move the last line of text inside the snippet as a comment.

Both cases would benefit from it, and we could share the same snippet.

See what you think.

suggestion

Comment (arboleya) [reply apps/docs/src/guide/predicates/methods.md]: Ooof, I didn't see your message before sending mine.

The UI may not refresh while one is writing a reply. 😓

Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: > Ooof, I didn't see your message before sending mine.
> 
> The UI may not refresh while one is writing a reply. 😓

Happens to me quite frequently as well, I end up missing comments at times 😢 




> TBH, the text preceding this snippet could be clearer as well. 🙂
> 
> We could move the last line of text inside the snippet as a comment.
> 
> Both cases would benefit from it, and we could share the same snippet.
> 
> See what you think.
> 

I agree, and it promotes code reuse so I will make that change.

Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: See [`80af938` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80af938e61d654730c281f02f729725b79f97b10)

Comment (arboleya) [reply apps/docs/src/guide/predicates/methods.md]: Nice! Not sure, but then I guess you'll need to undo [this](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/2a8de44dc130234cfff5cc6b455c9ad6f1be188f#diff-bf9cefcebbc4d0127bda5f3a965bcb75c3d6e4130467325e4701ed38d0b91dd1R122-R156) as well I?

Anyway, I am leaving this to you and resolving it for now.

Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md]: Correct 😅

**Review by arboleya — COMMENTED** [review]
You have two unresolved conversations.

unresolved

File: apps/docs/src/guide/predicates/methods.md [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+ - `Transfers`
+
+
+## Checking Balances
+
+### `getBalances`
+
+This will return the balances of all assets owned by the predicate.
+
+See also: [Checking Wallet Balances](https://docs.fuel.network/docs/fuels-ts/wallets/checking-balances/#getting-a-wallets-balance)
```
Comment (arboleya) [inline apps/docs/src/guide/predicates/methods.md:15]: I realized my first comment may have been misleading.

What I meant was to remove this section entirely, and then — see the next suggestion.

```suggestion
```

File: apps/docs/src/guide/predicates/methods.md [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+### `getResourcesToSpend`
+
+This will return the resources owned by a predicate so that they can be added to a transaction request.
+
+This method is called under the hood when using [`transfer`](./methods.md#transfer) or [`createTransfer`](./methods.md#createtransfer).
+
+You may want to use this method when using a predicate in an existing transaction request.
+
+<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-1{ts:line-numbers}
+
```
Comment (arboleya) [inline apps/docs/src/guide/predicates/methods.md:26]: Then the "see also" would go below this snippet here.

However, considering there is no `getBalances` for `Predicates`, why should we have this here?

```suggestion

See also:
 - [Checking Wallet Balances](https://docs.fuel.network/docs/fuels-ts/wallets/checking-balances/#getting-a-wallets-balance)
```

**Review by arboleya — APPROVED** [review]

File: apps/docs/src/guide/predicates/methods.md (hunk 26 above)
Comment (arboleya) [reply apps/docs/src/guide/predicates/methods.md:15]: Dismissing my own suggestion. 🤦‍♂️

File: apps/docs/src/guide/predicates/methods.md (hunk 27 above)
Comment (arboleya) [reply apps/docs/src/guide/predicates/methods.md:26]: Never mind.

Since `Predicates` extends from `Account`, there is also a `getBalances` for it (thanks, @Torres-ssf, for the heads-up).

With that in mind, forget everything I said. 🤦‍♂️ You could leave it as is or roll it back to your first version, as that would be more complete and remove the need for users to click around.

Perhaps the phrasing should target `Account` instead of `Wallets`, though. I'm unsure if that makes sense, but it can also be a follow-up. Sorry for the confusion.

Comment (maschad) [reply apps/docs/src/guide/predicates/methods.md:26]: No worries, it was also a bit misleading for me before I understood that a predicate is a type of account at first. I think the current state makes sense and still references another example of getting balances that consumers may also be familiar with.

**Review by nedsalk — DISMISSED** [review]

**Review by Torres-ssf — APPROVED** [review]

**Review by arboleya — APPROVED** [review]
